### ISSUE TITLE
Don't limit /dev/shm to 1G

### ISSUE THREAD

**Issue by strugee** [QubesOS/qubes-issues#5300]
**The problem you're addressing (if any)**

`/dev/shm` is too small. On a regular Linux system `/dev/shm` is supposed to be half the size of physical RAM but in Qubes - at least in the Debian template - it is limited to 1G in `/etc/fstab`.

**Describe the solution you'd like**

I think probably we should mount `/dev/shm` as half the maximum amount of RAM the qube is allowed to use. That seems to best match what baremetal systems do. Note also that tmpfs only consumes RAM when written to so increasing this value will not increase memory usage unless applications actually request/use that amount of memory.

**Where is the value to a user, and who might that user be?**

I have a very large number of Firefox tabs and experience extremely frequent crashes which I have tracked back to `/dev/shm` being too small. See https://bugzilla.mozilla.org/show_bug.cgi?id=1245239. The value to the user is better system reliability because they will be less likely to trigger bugs like this one.

**Describe alternatives you've considered**

None that I can see. We could set it to a fixed value that's higher than the current 1G, but we'll never get it right for _all_ qubes. What if the qube is super minimal and doesn't actually have that amount of RAM available? Or, what if it's huge and the value we pick is still way too small?

**Relevant [documentation](https://www.qubes-os.org/doc/) you've consulted**

None

**Related, [non-duplicate](https://www.qubes-os.org/doc/reporting-bugs/#new-issues-should-not-be-duplicates-of-existing-issues) issues**

 #1003 (IIRC Firefox uses the same IPC implementation code as Chrome)

**Comment by marmarek** [comment]
This may be a good idea. The current 1GB limit is because by default Linux use half of available memory **at system start time**. Which means 200MB on Qubes (given default 400M initial memory). This is clearly too little.
I'm not yet sure what is the easiest way to set this limit dynamically at boot time. Remount from one of a startup scripts? Generate systemd .mount unit dynamically?
One of the methods to get maxmem from within qube is `xenstore-read memory/static-max`.

**Comment by strugee** [comment]
Right, I figured it was something like that. Maybe we can adjust fstab early in bootup, similar to how we resize the rootfs before anything else comes up?

**Comment by marmarek** [comment]
I'd avoid dynamic config file modifications - this is very fragile and should be used as a last resort. That's why I'm considering systemd .mount units - you can modify it by putting a separate file in `dev-shm.mount.d` directory. But I haven't verified if you can override just one option this way.

**Comment by strugee** [comment]
We could also use a systemd generator.

**Comment by brendanhoar** [comment]
How and when will the "reevaluated" value be generated before being applied to the VM? 

Will it be passed from dom0? 

If so, could this possibly be a value that can be overridden/customize via qubes-preference settings? And if so, by value, percent of maxmem or both?

**Comment by strugee** [comment]
Update: I asked about this on systemd IRC, because `dev-shm.mount` doesn't exist on my system and I found that confusing. Apparently the reason is that API filesystems (including `/dev/shm`) are in a hard-coded list in pid 1 that are mounted very early at boot time. I was told that `systemd-remount-fs.service` will remount API filesystems listed in `fstab`, but that it _only_ considers `fstab`. In other words, if we create `dev-shm.mount`, it will be a noop. I am guessing based on [this comment from Lennart](https://github.com/systemd/systemd/issues/8394#issuecomment-374698015), but have not confirmed, that even if we create `dev-shm.mount`, do `systemctl daemon-reload` and then restart `dev-shm.mount`, it still will not work. (Even if it did, upstream would still consider it Wrong and might break us.)

That leaves us with either modifying `fstab`, or doing `mount -o remount` at runtime. The latter doesn't actually seem too bad to me; we could just wrap it in a service unit - `qubes-remount-dev-shm.service` or something like that. We could have it bail out if an entry exists in `fstab`, which would allow users to override the default.

> If so, could this possibly be a value that can be overridden/customize via qubes-preference settings? And if so, by value, percent of maxmem or both?

@brendanhoar what would the usecase for this be? Seems like a pretty advanced setting that most people would have no reason to change. Remember, this filesystem won't actually consume more RAM unless it's used by an application. It's just tmpfs under the hood.
